Add documentation (#87)
precice/systemtests: docs/adapters-ci.md

> # General architecture 

Reviewer: Don't we need a different title here? This is the same as in `main.md`.

> 

Reviewer: Maybe add a few introductory lines here? Currently the sections feel a bit disconnected in first sight.

## Triggering the build from the adapter 

> In order not to duplicate tests on different adapters, and reuse functionality provided here, Travis API 

Reviewer: Active voice is easier to read in this case:
```suggestion
In order not to duplicate tests on different adapters, and to reuse functionality provided here, we use the Travis API 
```

> is used to trigger selected tests from this repository after commits to the adapters repositories. This is done by

Reviewer: ```suggestion
to trigger selected tests from this repository after commits to the adapters repositories. This is done by
```

> generating custom JSON with specification of the job and sending it to the Travis that triggers the build (For overview of build triggering see [here](https://docs.travis-ci.com/user/triggering-builds/).

Reviewer: * What is a "JSON with specification of the job"?
* Which is the "Travis (job?) that triggers the build"? Which other Travis ?jobs? do we have? Would be good to document this.

Author: > JSON with specification of the job

This is explained in the provided link to Travis API

> Which is the "Travis (job?) that triggers the build"? Which other Travis ?jobs? do we have? Would be good to document this.

I will elaborate a bit more on this.

> generating custom JSON with specification of the job and sending it to the Travis that triggers the build (For overview of build triggering see [here](https://docs.travis-ci.com/user/triggering-builds/).

Reviewer: ```suggestion
generating custom JSON with specification of the job and sending it to the Travis that triggers the build (For overview of build triggering see [here](https://docs.travis-ci.com/user/triggering-builds/)).
```

> Once the tests are finished their return status is queried/sent back (controlled by `--wait` argument) to the Travis job on the adapter repository, making it fail/pass, depending on the return status.

Reviewer: ```suggestion
Once the tests are finished, their return status is queried/sent back (controlled by `--wait` argument) to the Travis job on the adapter repository, making it fail/pass, depending on the return status.
```

> Once the tests are finished their return status is queried/sent back (controlled by `--wait` argument) to the Travis job on the adapter repository, making it fail/pass, depending on the return status.

Reviewer: "queried" and "sent back" sound like different actions with different subjects here:

* queried: the caller asks the callee (e.g. every some time)
* send back: the callee returns after finishing its task

Which one do we have here?

Author: We have both . It is controlled by `--wait` argument. In one case the job on the adapter Travis query the status (every some time), so it is kind of blocking call if you run with `--wait` flag. Without it it is non blocking, it just triggers the build and returns. If systemtests fails, the return status is sent back, causing to generated "Failed job" on the adapter. I might elaborate more on this in the documentation.

> You can learn more from `trigger_systemtests.py`, that handles all job generation and triggering.

Reviewer: ```suggestion
You can learn more from `trigger_systemtests.py`, which handles all job generation and triggering.
```

> Since we don't need run full set of tests, when generating JSON only following stages are added:

Reviewer: ```suggestion
Since we don't need to run a full set of tests, only the following stages are added when generating the JSON job configuration file:
```

- Building the adapter 
- Running the tests in which it participates

## Running on different branches and pull requests

> During the Travis job on the adapter if it is identified, that the job is running on pull request on non-master branch (using 

Reviewer: ```suggestion
During the adapter's Travis job, if it is identified that the job is running on a pull request on a non-master branch (using the 
```

> `TRAVIS_BRANCH` and `TRAVIS_PULL_REQUEST` environment variables), the JSON is patched with additional bash script that modifies

Reviewer: (is actually the JSON file patched with a script?)
```suggestion
`TRAVIS_BRANCH` and `TRAVIS_PULL_REQUEST` environment variables), the JSON file is patched with an additional bash script that modifies
```

Author: Yep...This way prior to building, the script is executed and  dockerfiles get additional commands as wel.

Reviewer: Before I understood "we add a script to the JSON file", but I see that you actually mean "we use a script to patch the JSON files". Maybe naming the script would avoid the confusion here (although I agree, that was not a very bright thought).

> the adapter Dockerfiles with additional instruction to switch to the needed branch (obviously the instruction is injected prior 

Reviewer: I would avoid using terms such as "obviously" in documentation, as they are emotionally charged. The term "unsurprisingly" conveys the same message in a more neutral way. However, I would still just skip the qualifier.
```suggestion
the adapter Dockerfiles with additional instruction to switch to the requested branch (prior 
```

to actually building an image).

> The resultant image is then tagged with the name of the branch/pull request and pushed to the Dockerhub. In the `docker-compose.yml` the image tag 

Reviewer: A link would be useful:
```suggestion
The resultant image is then tagged with the name of the branch/pull request and pushed to [Dockerhub](https://hub.docker.com/u/precice). In `docker-compose.yml`, the image tag 
```

> is controlled using  `{${SOLVER_NAME}_TAG}`( defaults to `latest`). Similarly this environment variables is also specified in the specification of the Travis job

Reviewer: ```suggestion
is controlled using  `{${SOLVER_NAME}_TAG}`( defaults to `latest`). Similarly, this environment variable is also specified in the specification of the Travis job
```

for every solver participating in the test.

> ## Adding CI on the adapter

Reviewer: ```suggestion
## Adding CI for an adapter
```

> Add specifications of the adapter to the `adapter_info` dictionary in `trigger_systemtests`.

Reviewer: ```suggestion
Add specifications of the adapter into the `adapter_info` dictionary in `trigger_systemtests.py`.
```

> Then the only thing remaining is to add `.travis.yml` to the adapter:

Reviewer: ```suggestion
Then, the only thing remaining is to add a `.travis.yml` file to the adapter's repository, such as:
```

    - if: branch = master
      script: |
        curl -LO --retry 3 https://raw.githubusercontent.com/precice/systemtests/master/trigger_systemtests.py
        python3 trigger_systemtests.py --adapter openfoam

> We also need to add `TRAVIS_ACCESS_TOKEN` in each adapter repository as a secure variable to authenticate build triggering. It can be found on the [Travis profile page](https://travis-ci.org/account/preferences).

Reviewer: ```suggestion
We also need to add a `TRAVIS_ACCESS_TOKEN` in each adapter's repository, as a secure variable to authenticate the build triggering. This token can be found on the [Travis profile page](https://travis-ci.org/account/preferences).
```

> We also need to add `TRAVIS_ACCESS_TOKEN` in each adapter repository as a secure variable to authenticate build triggering. It can be found on the [Travis profile page](https://travis-ci.org/account/preferences).

Reviewer: We also need a note like 

>At the end, remember to document the new test in the README.md and docs/main.md.

Author: > At  the end, remember to document the new test in the README.md and docs/main.md. 

This should be added, but not here probably. Here I describe how to add CI on adapters-repository, given that tests already exists. I should this comment in `main.md` or `development`.

will be then seen in Travis dashboard - one for systemtests and one for the adapter). Alternatively, non-blocking call is possible - in that case, in order to indicate
failure of the test, the script will trigger "failing job" on the adapters repository if the test fails.

You can learn more from `trigger_systemtests.py`, which handles all of described functionality.

Since we don't need to run a full set of tests, only the following stages are added when generating the JSON job configuration file:
- Building the adapter
- Running the tests in which it participates

## Running on different branches and pull requests

During the adapter's Travis job, if it is identified that the job is running on a pull request on a non-master branch (using the
`TRAVIS_BRANCH` and `TRAVIS_PULL_REQUEST` environment variables), the generated JSON filed is patched with additional bash script that modifies
the adapter Dockerfiles with additional instruction to switch to the needed branch (the instruction should be injected prior
to actually building an image).

> The resultant image is then tagged with the name of the branch/pull request and pushed to [Dockerhub](https://hub.docker.com/u/precice). In the `docker-compose.yml` the image tag

Reviewer 2: ```suggestion
The resulting image is then tagged with the name of the branch/pull request and pushed to [Dockerhub](https://hub.docker.com/u/precice). In the `docker-compose.yml` the image tag
```
